rust/tableau_summary/src/twb/raw/datasource/connection.rs

```diff
@@ -0,0 +1,489 @@
+                .log_error("ordinal not a number")
+                .unwrap_or_default())
+            .unwrap_or_default();
+        Self {
+            name: n.get_attr(NAME_KEY),
+            datatype: n.get_attr("datatype"),
+            ordinal,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, PartialEq, Clone, Debug)]
+pub struct ColMapping {
+    pub cols: HashMap<String, (String, String)>,
+}
+
+impl<'a, 'b> From<Node<'a, 'b>> for ColMapping {
+    fn from(n: Node) -> Self {
+        if n.get_tag() != "cols" {
+            info!("trying to convert a ({}) to cols", n.get_tag());
+            return Self::default();
+        }
+        let cols = n.find_tagged_children("map")
+            .into_iter()
+            .flat_map(map_node_to_kv)
+            .collect();
+        Self {
+            cols,
+        }
+    }
+}
+
+fn map_node_to_kv(n: Node) -> Option<(String, (String, String))> {
+    let key = n.get_attr("key");
+    let val = n.get_attr("value");
+    let val_parts = parse_identifiers(&val);
+    let val_tuple = if let Some((x, y)) = val_parts.into_iter().collect_tuple() {
+        (x, y)
+    } else {
+        info!("found col value: {val} that didn't parse into 2 parts");
+        return None;
+    };
+    Some((key, val_tuple))
+}
+
+// Given a string like `[foo].[bar]` return ["foo", "bar"]
+// TODO: see what tds / twb files do when the table or column has a `.` in the name
+//       currently, we just assume it is not escaped or anything
+//       (e.g. "[foo.1].[bar.txt]" -> ["foo.1", "bar.txt"])
+pub fn parse_identifiers(s: &str) -> Vec<String> {
+    let mut result = Vec::new();
+    let mut current_entry = String::new();
+    let mut inside_brackets = false;
+
+    for c in s.chars() {
+        match c {
+            '[' => {
+                inside_brackets = true;
+            }
+            ']' => {
+                inside_brackets = false;
+                result.push(current_entry.clone());
+                current_entry.clear();
+            }
+            '.' if inside_brackets => {
+                current_entry.push('.');
+            }
+            '.' if !inside_brackets => {},
+            _ => {
+                current_entry.push(c);
+            }
+        }
+    }
+
+    result
+
+}
+
+#[derive(Serialize, Deserialize, Default, PartialEq, Clone, Debug)]
+pub struct MetadataRecords {
+    // table -> aggregation
+    pub capabilities: HashMap<String, String>,
+    // name -> metadata
+    pub columns: HashMap<String, ColumnMetadata>,
+}
+
+impl<'a, 'b> From<Node<'a, 'b>> for MetadataRecords {
+    fn from(n: Node) -> Self {
+        if n.get_tag() != "metadata-records" {
+            info!("trying to convert a ({}) to metadata-records", n.get_tag());
+            return Self::default();
+        }
+        let mut capabilities = HashMap::new();
+        let mut columns = HashMap::new();
+        n.find_tagged_children("metadata-record")
+            .into_iter()
+            .for_each(|c| {
```

Comment:
`for c in n.find_tagged_children("metadata-record") {`
you can deindent a bit and also don't need a separate function call per item (although I assume this is inlined by the compiler)